Refactor error handling page (#649)
stellar/stellar-docs: network/horizon/admin-guide/error-handling.mdx

---
title: Error Handling

> ---

Reviewer: Yeah I like API Reference -> Errors

> ---

Author: Yeah makes sense.

<CodeExample>

    ```bash
    curl -s https://horizon-testnet.stellar.org/claimable_balances/0000 | jq '.extras'
    {
        "invalid_field": "id",
        "reason": "Invalid claimable balance ID"
    }
    ```

</CodeExample>

Note that the SDKs make it a point to distinguish an invalid request (as above) versus a missing resource (a `404 Not Found`) (for example, the generic `NetworkError` versus a `NotFoundError` in the JavaScript SDK), where the latter might not be considered an error depending on your situation.

## Error Handling for Transaction Submissions

Horizon currently supports two types of transaction submission endpoints:

1. `/transactions_async`: Horizon submits a transaction to core in an async manner and passes the relevant core response back to the client. It is then the client's responsibility to poll for the status of that transaction.

> 2. `/transactions`: Horizon submits a transaction to core and then waits for it to be ingested in the database. It will either timeout while waiting or return a transaction response on success.

Author: @Shaptic It is `/transactions`: https://developers.stellar.org/network/horizon/api-reference/resources/transactions

There are some resolution strategies that are common between the 2 endpoints while some strategies are more endpoint specific.

### Request Adjustments

Certain transaction submission failures also need adjustments to succeed.

- If the XDR is malformed, or the transaction is otherwise invalid, you’ll encounter a `400 Bad Request` (for example, see excluding a source account). Both transactions and their operations can be easily malformed: look at the `extras.result_codes` field for details and cross-reference them with the appropriate result codes documentation to determine specifics.
- Transaction fees are also a safe adjustment. If you get a `tx_insufficient_fee` error, refer to the Insufficient Fees and Surge Pricing section later in this document.

### Polling and Retrying Transactions

#### Async Transaction Submission

> Submissions using the async endpoint return a clear response back from core. There are different actions that clients can take based on the specific `tx_status` returned:

Reviewer: ```suggestion
Submissions using the `/transactions_async` endpoint return an immediate response back from core. There are different actions that clients can take based on the specific `tx_status` returned:
```

> 1. `PENDING`: the submission is successful but the txn is still waiting to be included in the ledger. You should use the `GET /transactions/:hash` endpoint to poll for the submitted txn and check if it makes into the ledger.

Author: @Shaptic Good idea. I will keep in mind for future doc changes

> 2. `ERROR`: the submission did not go through due to some error with core. Take a look at the attached error message and retry the submission again.

Reviewer: ```suggestion
2. `ERROR`: the submission did not go through due to an error in Stellar Core. Take a look at the attached error message for more details, modify your transaction if necessary, and resubmit.
```

> 2. `ERROR`: the submission did not go through due to some error with core. Take a look at the attached error message and retry the submission again.

Author: @JakeUrban George is correct. The `ERROR` specifically indicates the submission did not go through unlike the ambiguity behind `504 Timeout`. So in this case the user has to retry the submission again.

> 2. `ERROR`: the submission did not go through due to some error with core. Take a look at the attached error message and retry the submission again.

Reviewer: I started making some of these suggestions as well but going to stop now. I recommend just doing a hollistic pass through this for these kinds of things though. That is, removing abbreviations and shorthands

> 3. `TRY_AGAIN_LATER`: this indicates that the stellar-core instance is currently unable to process any submission requests. Clients should wait for sometime before retrying it again.

Author: @JakeUrban I referenced this: https://github.com/stellar/stellar-core/blob/b3aeb14cc798f6d11deb2be913041be916f3b0cc/src/simulation/LoadGenerator.h#L272: for writing it but looks like you and George are correct here.

#### Synchronous Transaction Submission

> Due to the blocking nature of the non-async endpoint, things are a little different compared to the async strategy. There are 3 possible scenarios that clients can encounter:

Reviewer: ```suggestion
Due to the blocking nature of this endpoint, things are a little different compared to the asynchronous strategy. There are 3 possible scenarios that clients can encounter:
```

1. The submission is successful and Horizon returns the transaction response back. This is the happy path and clients do not need to do anything but wait for Horizon's response.

> 2. Core sends back an `ERROR` response from the submission. Clients should consult the attached error message and retry the submission again.

Reviewer: They might, right? I think there are some responses that core will send back immediately on submission attempt (which Horizon propagates through), but I might be mistaken

> 2. Core sends back an `ERROR` response from the submission. Clients should consult the attached error message and retry the submission again.

Reviewer 2: Yes thats true (such as bad sequence number) but I guess my point is that Horizon returns its own error response, not the `ERROR` response from core.

> 2. Core sends back an `ERROR` response from the submission. Clients should consult the attached error message and retry the submission again.

Author: @JakeUrban Horizon does propogate the exact error back from core - https://github.com/stellar/go/blob/master/services/horizon/internal/txsub/submitter.go#L63

> 2. Core sends back an `ERROR` response from the submission. Clients should consult the attached error message and retry the submission again.

Author: @JakeUrban We are actually returning the error response from core here: https://github.com/stellar/go/blob/master/services/horizon/internal/txsub/submitter.go#L63.

> 3. Timeouts: Horizon may send a `504 Timeout` after transaction submission. Timeouts are not errors but warnings that your request hasn’t been fulfilled yet. This can happen because of the relationship between Horizon and Stellar Core- the network may take some time (5-10 mins during congestion) to accept the transaction. At the same time, Horizon needs to provide developers with a response within 30 seconds. Due to the blocking nature of the endpoint, there could be many possible scenarios behind the timeout - transient outages, congestion on the network, issues on Core side, issues on Horizon side etc.... There is no way to cancel a transaction after submitting it. So, you need to carefully consider before resubmitting your transactions.

Reviewer 2: ```suggestion
3. Timeouts: Horizon may respond with a `504` HTTP code. This response is not an error but a warning that your transaction hasn't been accepted by the network yet. There could be many possible scenarios behind the timeout, the most common of which is network congestion, but it could also be due to other transient issues. There are a couple things you should to do best handle this response.
```

   1. `Polling the transaction hash`: The recommended step is to use the txn hash in the timeout response and poll the `GET /transactions` endpoint to see if it successfully made it into a ledger.

>    2. `Time bounds`: Time bounds are optional but recommended, as they put a time limit on the transaction - so either the transaction makes it onto the ledger or it times out depending on your time parameters.

Reviewer: This doesn't feel like it belongs as a bullet point of `There are a couple things you should to do best handle this response.`. It's more like a preventative measure. I would probably just rephrase that leading sentence, rather than changing this list though

>    3. `Resubmitting the transaction`: If the txn status poll has not returned anything or if your transaction has timed out based on the time bounds you specified, you should attempt a resubmission of the same transaction. If the transaction has already successfully made it to the ledger, Horizon will return any attempted resubmission. Only in cases where a transaction’s status is unknown (and thus will have a chance to make it on the ledger) will a resubmission to the network occur.

Reviewer 2: I don't think we want to mention this, as it contracts the suggestion to poll using the transaction hash.

Reviewer 3: Agreed, we should discourage relying on Horizon's unnecessary helpfulness in this case

Horizon will return a valid txn response for any resubmissions for successful transactions. However, do note that resubmitting a txn is only safe when it is unchanged - same operations, signatures, sequence number, etc.... Be careful when working around an error that does require changes to the transaction. It can cause duplicate transactions, which can cause problems - double payments, incorrect trustlines, and more. If you continue to face timeouts on retries, consider using a fee-bump transaction to get into the ledger (after the time bounds expire) or increasing the maximum fee you’re willing to pay. Read up on [Surge Pricing and Fee Strategies](docs/learn/fundamentals/fees-resource-limits-metering) for more details.

> ### Example: Using Time Bounds

Author: Hmm it can be used both ways. I think writing it separately looks good here.

We assume the existence of a sleep implementation similar to the one [here](https://stackoverflow.com/a/39914235) Be sure to integrate backoff into your retry mechanism. In our example error-handling code above, we implement a simple linear backoff, but there are [plenty of recommendations](https://backoff-utils.readthedocs.io/en/latest/strategies.html#why-are-backoff-strategies-useful) for various other strategies. Backoff is important both for maintaining performance and avoiding rate-limiting issues.

### Example: Invalid Sequence Numbers

These errors typically occur when you have an outdated view of an account. This could be because multiple devices are using this account, you have concurrent submissions happening, or other reasons. The solution is relatively simple: retrieve the account details and try again with an updated sequence number.

<CodeExample>

```js
// suppose `account` is an outdated `AccountResponse` object
let tx = sdk.TransactionBuilder(account, ...)/* etc */.build()
server.submitTransaction(tx).catch(function (error)) {
  if (error.response && error.status == 400 && error.extras &&
      error.extras.result_codes.transaction == sdk.TX_BAD_SEQ) {

>     return server.accounts()
>       .accountId(account.accountId())
>       .then(function (response) {

Reviewer 3: ```suggestion
    return server.loadAccount(account.accountId())
      .then(function (response) {
```

        let tx = sdk.TransactionBuilder(response, ...)/* etc */.build()
        return server.submitTransaction(tx);
      });
  }
  // ...other error conditions...
}
```

</CodeExample>

Despite the solution’s simplicity, things can go wrong fast if you don’t understand why the error occurred.

> Suppose you submit transactions from multiple places in your application simultaneously, and your user spammed the Send Payment button a few times in their impatience. If you send the exact same payment transaction for each tap, naturally, only one will succeed. The others will fail with an invalid sequence number (`tx_bad_seq`), and if you resubmit blindly with an updated sequence number (as we do above), these payments will also succeed, resulting in more than one payment being made when only one was intended. So, be very careful when resubmitting transactions that have been modified to work around an error.

Reviewer 3: Suggestion: emphasis 
```suggestion
Suppose you submit transactions from multiple places in your application simultaneously, and your user spammed the Send Payment button a few times in their impatience. If you send the exact same payment transaction for each tap, naturally, only one will succeed. The others will fail with an invalid sequence number (`tx_bad_seq`), and if you resubmit blindly with an updated sequence number (as we do above), these payments will also succeed, resulting in more than one payment being made when only one was intended. So **be very careful when resubmitting transactions** that have been modified to work around an error.
```

| `INSUFFICIENT_BALANCE`  | -7   | Fee would bring account below minimum balance; see our section on [Lumens](/docs/learn/fundamentals/lumens#minimum-balance) for more info |
| `NO_ACCOUNT`            | -8   | Source account not found                                                                                                                  |
| `INSUFFICIENT_FEE`      | -9   | Fee is too small; see our section on [Fees](/docs/learn/fundamentals/fees-resource-limits-metering) for more info                         |
| `BAD_AUTH_EXTRA`        | -10  | Unused signatures attached to transaction                                                                                                 |
| `INTERNAL_ERROR`        | -11  | An unknown error occurred                                                                                                                 |
| `NOT_SUPPORTED`         | -12  | The transaction type is not supported                                                                                                     |

> | `FEE_BUMP_INNER_FAILED` | -13  | The fee bump inner transaction failed                                                                                                     |

Reviewer 3: Suggestion: clarity

```suggestion
| `FEE_BUMP_INNER_FAILED` | -13  | The inner transaction of a fee-bump transaction failed                                                                                                     |
```

| `BAD_SPONSORSHIP`       | -14  | The sponsorship is not confirmed                                                                                                          |

### Insufficient fees and surge pricing

See the [Fees section](/docs/learn/fundamentals/fees-resource-limits-metering)

### Rate limiting

> If you’re using SDF’s public Horizon instance, you may get a `429 Too Many Requests` error when exceeding the rate limits. If you’re encountering this frequently, it may be time to deploy your own Horizon instance!

Reviewer 3: Suggestion: cross-reference to the admin guide

> If you’re using SDF’s public Horizon instance, you may get a `429 Too Many Requests` error when exceeding the rate limits. If you’re encountering this frequently, it may be time to deploy your own Horizon instance!

Reviewer: We removed all references to SDF horizon URL in docs, I think I'd prefer to leave out the mention of it explicitly here as well.

```suggestion
Horizon may rate limit requests and return `429 Too Many Requests` error when exceeding the rate limits. If using your own instance, you may want to either increase or disable rate limiting. If you're using a third-party Horizon instance, you may want to deploy your own to have more control over this configuration.
```
